fix(lsp): add "silent" option to vim.lsp.start

gpanders opened the pull request:
vim.notify cannot be suppressed and it is not always necessary to display a visible warning to the user if the RPC process fails to start. For instance, a user may have the same LSP configuration across systems, some of which may not have all of the LSP server executables installed. In that case, the user receives a notification every time a file is opened that they cannot suppress.

Instead of using vim.notify in vim.lsp.rpc, propagate a normal error up through the call stack and use vim.notify in vim.lsp.start() only if the "silent" option is not set.

This also updates lsp.start_client() to return an error message as its second return value if an error occurred, rather than calling vim.notify directly. Callers of lsp.start_client() will need to update call sites appropriately if they wish to report errors to the user (or even better, switch to vim.lsp.start).

gpanders committed:
fix(lsp): add "silent" option to vim.lsp.start

vim.notify cannot be suppressed and it is not always necessary to
display a visible warning to the user if the RPC process fails to start.
For instance, a user may have the same LSP configuration across systems,
some of which may not have all of the LSP server executables installed.
In that case, the user receives a notification every time a file is
opened that they cannot suppress.

Instead of using vim.notify in vim.lsp.rpc, propagate a normal error up
through the call stack and use vim.notify in vim.lsp.start() only if
the "silent" option is not set.

This also updates lsp.start_client() to return an error message as its
second return value if an error occurred, rather than calling vim.notify
directly. Callers of lsp.start_client() will need to update call sites
appropriately if they wish to report errors to the user (or even better,
switch to vim.lsp.start).

MariaSolOs commented:
> vim.notify cannot be surpressed and it is not always necessary to display a visible warning to the user if the RPC process fails to start.

Although it might be noisy, I personally think that having a visible and urgent notification is suitable in this case, and that a message in the log file is easily lost.

Perhaps we can add an option to `vim.lsp.rpc.start` on what to do in these scenarios to please everyone?

mfussenegger commented on runtime/lua/vim/lsp.lua:
```diff
@@ -244,6 +244,16 @@ function lsp.start(config, opts)
   local bufnr = resolve_bufnr(opts.bufnr)
 
+  local fname = vim.api.nvim_buf_get_name(bufnr)
+  if
+    vim.bo[bufnr].buftype ~= ''
+    or not vim.api.nvim_buf_is_loaded(bufnr)
+    or not vim.uv.fs_access(fname, 'r')
+  then
+    -- Not a valid buffer
+    return
+  end
```
I'm not sure I understand the rationale for this?

Attaching lsp to buffers which are not backed by the filesystem is a legitimate use-case. This would break plugins like nvim-jdtls which support serving `jdt://` uri's via a custom off-spec extension:

https://github.com/mfussenegger/nvim-jdtls/blob/8eb5f0dbe6e126b392ddcaf45893358619893e45/plugin/nvim_jdtls.vim#L6-L7

https://github.com/mfussenegger/nvim-jdtls/blob/8eb5f0dbe6e126b392ddcaf45893358619893e45/lua/jdtls.lua#L1164-L1215

I think https://github.com/sourcegraph/sg.nvim uses a similar approach, and in the LSP repo there are discussions like this: https://github.com/microsoft/language-server-protocol/issues/1264

lewis6991 replied:
Yes I agree, I often open scratch buffers for Lua stuff that don't get written.

gpanders replied:
Ok this is a case of over fitting to my use case I guess. `clangd` does not work without a backing file so I (incorrectly) assumed that this was the case with all servers.

Will drop the `vim.uv.fs_access` check, but `nvim_buf_is_loaded` and `buftype ~= ''` are still ok I think? Or do either of you know of uses cases where those might be invalid assumptions?

lewis6991 replied:
`buftype` might be going too far. In my config I just have: `if vim.bo[args.buf].buftype == 'nofile' then`, but that may not be the correct thing to do generally.

mfussenegger replied:
> Will drop the vim.uv.fs_access check, but nvim_buf_is_loaded and buftype ~= '' are still ok I think? Or do either of you know of uses cases where those might be invalid assumptions?

I second that checking the `buftype` is still too much. Just to show another use-case: https://github.com/mfussenegger/dotfiles/blob/8f99373541a0928eaea7e8e6273307faa370cd34/vim/.config/nvim/commands.lua#L52-L95 there I use `acwrite` to intercept writes to execute code via `BufWriteCmd`. lua-language-server works fine on that scratch buffer.

I think only the `nvim_buf_is_loaded` check is fine, but that's also already present in a way: https://github.com/neovim/neovim/blob/a1550dbf0ac252052a7a71780708d3ec05aa3686/runtime/lua/vim/lsp.lua#L596-L599

lewis6991 replied:
What about disallowing `nofile`? I can't remember why I needed it, but it was to prevent attaching to some intermediate buffers, probably gitsigns diff buffers.

mfussenegger replied:
`nofile` definitely works. I use that in nvim-jdtls.

I think the only buftype that might cause problems is `prompt`, but that might also have been fixed with https://github.com/neovim/neovim/pull/28260

justinmk commented:
>  Since vim.lsp.start() is a slightly higher level interface over Client.start_client()

Does this imply that start_client will continue to be a public interface? I was hoping we could eliminate it.

gpanders commented:
> > vim.notify cannot be surpressed and it is not always necessary to display a visible warning to the user if the RPC process fails to start.
> 
> Although it might be noisy, I personally think that having a visible and urgent notification is suitable in this case, and that a message in the log file is easily lost.
> 
> Perhaps we can add an option to `vim.lsp.rpc.start` on what to do in these scenarios to please everyone?

So long as there is some kind of affordance to suppress the warning or error message, I'm fine with this solution. I do not want to be notified every time I open a Python file on a remote server that `pyright` is not installed. I _know_ it's not installed, that's not a warning or an error.

gpanders commented:
> > Since vim.lsp.start() is a slightly higher level interface over Client.start_client()
> 
> Does this imply that start_client will continue to be a public interface? I was hoping we could eliminate it.

I guess this depends on whether we want to support "advanced" LSP use cases for plugins, or if we want to have everyone just use `vim.lsp.start()`. I'm fine with the latter, personally, but I don't know if/how other plugins are using `start_client()` (I know nvim-lspconfig is, but that's only because nvim-lspconfig hasn't been meaningfully updated in a long time).

mfussenegger commented:
> So long as there is some kind of affordance to suppress the warning or error message, I'm fine with this solution. I do not want to be notified every time I open a Python file on a remote server that pyright is not installed. I know it's not installed, that's not a warning or an error.

I'd also prefer a flag to ignore these errors over removing the default notification.
For people initially setting up LSP it's not nice if there's no visible feedback at all if the process can't be found or fails to execute.

gpanders commented:
> > So long as there is some kind of affordance to suppress the warning or error message, I'm fine with this solution. I do not want to be notified every time I open a Python file on a remote server that pyright is not installed. I know it's not installed, that's not a warning or an error.
> 
> I'd also prefer a flag to ignore these errors over removing the default notification. For people initially setting up LSP it's not nice if there's no visible feedback at all if the process can't be found or fails to execute.

We can add a `quiet` field to the `opts` parameter of `vim.lsp.start()`.

Then in `rpc.start()` we can use `error()` instead of vim.notify. `vim.lsp.start()` uses `pcall` when calling `lsp.start_client` and calls `vim.notify` if `quiet` is falsey.


This will kind of break nvim-lspconfig though, since they still use `start_client` directly (users will see an error rather than a warning notification), though it should be an easy update on that side (eventually, nvim-lspconfig should use `vim.lsp.start()` too, but that's a bigger effort).

MariaSolOs commented on runtime/doc/lsp.txt:
```diff
@@ -2340,7 +2342,7 @@ start({cmd}, {dispatchers}, {extra_spawn_params})        *vim.lsp.rpc.start()*
 
     Return: ~
-        (`vim.lsp.rpc.PublicClient?`) Client RPC object, with these methods:
+        (`vim.lsp.rpc.PublicClient`) Client RPC object, with these methods:
```
Pardon my ignore here, but does this change in the return type mean that in order for a plugin to check if creating an LSP client succeeds, they can no longer rely on checking if this returns `nil` or not and instead have to wrap the call with a `pcatch`?

gpanders replied:
Yes. Need to do a proper code search, but as far as I’m aware no plugins are using this API directly except nvim-lspconfig, which I will update myself when this is merged.

Alternatively we could use multi return values, `client|nil, nil|string` where the 2nd argument is the error if the client failed to start. That’s still a breaking change, but slightly less breaking.

MariaSolOs replied:
Got it, thanks for the early research. I think the current implementation is good for now, but given that `vim.lsp.start` wasn't added in v0.10, the change at least deserves an entry in `news.txt`.

MariaSolOs reviewed (approved):
I just have a small question about the consequences of this change but otherwise it lgtm.

mfussenegger commented on runtime/lua/vim/lsp.lua:
```diff
@@ -244,17 +247,29 @@ function lsp.start(config, opts)
   local reuse_client = opts.reuse_client or reuse_client_default
   local bufnr = resolve_bufnr(opts.bufnr)
 
+  if vim.bo[bufnr].buftype == 'nofile' or not vim.api.nvim_buf_is_loaded(bufnr) then
```
```suggestion
  if not vim.api.nvim_buf_is_loaded(bufnr) then
```

As mentioned in the comments - there are currently working `nofile` use-cases

gpanders replied:
Ah I completely misinterpreted your [comment](https://github.com/neovim/neovim/pull/28478#discussion_r1578000992) as "`nofile` definitely works [as the value to use in this PR]". My mistake.

gpanders replied:
Ok given that `'buftype'` is not in use at all I'm just going to drop the first commit and make this PR only adding the `silent` option (since `nvim_buf_is_loaded` is already checked in `buf_attach_client`).

justinmk commented on runtime/doc/lsp.txt:
```diff
@@ -869,12 +869,14 @@ start({config}, {opts})                                      *vim.lsp.start()*
                     re-uses a client if name and root_dir matches.
                   • {bufnr} (`integer`) Buffer handle to attach to if starting
                     or re-using a client (0 for current).
+                  • {quiet} (`boolean`) Suppress error reporting if the LSP
+                    server fails to start (default false).
```
`git grep quiet -- ':/*.lua' suggests that we haven't used this name anywhere yet. `silent` would be the conventional name.
```suggestion
                  • {silent} (`boolean`) Suppress error reporting if the LSP
                    server fails to start (default false).
```

gpanders commented:
>Alternatively we could use multi return values, client|nil, nil|string where the 2nd argument is the error if the client failed to start. That’s still a breaking change, but slightly less breaking.

I updated `vim.lsp.start_client()` to do this. I also updated the dispatch functions to return a client or error, rather than an optional client. This simplifies our error reporting paths.

This is a _mildly_ breaking change for anyone using `vim.lsp.start_client()` directly. Things will still work when they work, but if an error occurs there is no longer a notification by default. Callers of `start_client` need to handle the error themselves (or better yet, switch to `vim.lsp.start()`).

justinmk commented on runtime/doc/lsp.txt:
```diff
@@ -886,10 +888,11 @@ start_client({config})                                *vim.lsp.start_client()*
                   |vim.lsp.ClientConfig|.
 
-    Return: ~
+    Return (multiple): ~
         (`integer?`) client_id |vim.lsp.get_client_by_id()| Note: client may
         not be fully initialized. Use `on_init` to do any actions once the
         client has been initialized.
+        (`string?`) Error message, if any
```
Ideally we should follow the `pcall` convention of `boolean, string|result` but it's probably too breaking to do that here?

mfussenegger replied:
Not sure if it's too breaking, but I'd like to point out that `vim.uv` uses this pattern of `result|nil, errormsg|nil` for most of its functions. So in a way the pattern here wouldn't be new. Probably rather the opposite, there are fewer cases where `boolean, string|result` is used.

gpanders replied:
The Lua stdlib does this too, e.g. `io.open` returns `file|nil, nil|error`.

mfussenegger replied:
There is also an advantage to the `result|nil, errormsg|nil` pattern:
If you do:

```
local result, err = foobar()
if err then
    ...
elseif result then
    ...
end
```

luals can correctly infer the type of `result`.
With `bool, result|err` a `if ok` won't tell it if the second return value has the type of the error, or the type of the result.

justinmk replied:
interesting. let's document this pattern in `:help dev-lua` "interface conventions" section.

I should probably rework https://github.com/neovim/neovim/pull/28502 ?

gpanders commented:
nvim-lspconfig PR is here: https://github.com/neovim/nvim-lspconfig/pull/3126
